### PR TITLE
Add flake8 checking to CI tests

### PR DESCRIPTION
**Pull request by peterbarker**
This adds a step in our CI testing that checks Python files pass flake8 tests.

Only files including a special `AP_FLAKE8_CLEAN` marker are checked for cleanliness.

I've added `AP_FLAKE8_CLEAN` to the new CI script itself(!), and to `autotest.py` after tidying it up (again).  I expect more to grow this marker as they're cleaned up.

I've also added a script here which can be used as a git pre-commit hook.  It checks `.py` files which are staged (for either create or modify) with flake8 and fails the commit if they are not compliant.  This can stop bad Python getting into your commits in the first place.

Note that we already have some flake8 tests (supposedly?) running on Semaphore, so the use of the pre-commit hook could already save aggravation.

### CODE REVIEW

File: Tools/autotest/common.py [hunk 1]
```diff
@@ -2181,16 +2231,15 @@ def do_heartbeats(self, force=False):
...
+                now_wc - self.last_heartbeat_time_wc_s > 1):
             if x is not None:
                 self.last_heartbeat_time_ms = x.time_boot_ms
-            self.last_heartbeat_time_wc_s = now_wc
-            self.mav.mav.heartbeat_send(mavutil.mavlink.MAV_TYPE_GCS,
-                                        mavutil.mavlink.MAV_AUTOPILOT_INVALID,
-                                        0,
-                                        0,
-                                        0)
+                self.last_heartbeat_time_wc_s = now_wc
```
Comment (rmackay9) [inline Tools/autotest/common.py:2237]: this looks like a real bug fix which seems good.

File: Tools/autotest/common.py [hunk 2]
```diff
@@ -483,28 +515,30 @@ def write_data_id(self, dataid):
...
-        self.dataid_GPS_COURS_BP        = 0x14
-        self.dataid_GPS_SPEED_AP        = 0x19
-        self.dataid_GPS_LONG_AP         = 0x1A
-        self.dataid_GPS_LAT_AP          = 0x1B
-        self.dataid_BARO_ALT_AP         = 0x21
-        self.dataid_GPS_LONG_EW         = 0x22
-        self.dataid_GPS_LAT_NS          = 0x23
-        self.dataid_CURRENT             = 0x28
-        self.dataid_VFAS                = 0x39
+        self.dataid_GPS_ALT_BP          = 0x01  # noqa
```
Comment (tridge) [inline Tools/autotest/common.py]: i don't think we want this

File: Tools/autotest/common.py [hunk 3]
```diff
@@ -790,22 +829,22 @@ def __init__(self, destination_address, verbose=True):
...
-        self.SPORT_FRAME_XOR  = 0x20
+        self.SPORT_FRAME_XOR = 0x20
 
-        self.SENSOR_ID_VARIO             = 0x00 # Sensor ID  0
-        self.SENSOR_ID_FAS               = 0x22 # Sensor ID  2
-        self.SENSOR_ID_GPS               = 0x83 # Sensor ID  3
-        self.SENSOR_ID_RPM               = 0xE4 # Sensor ID  4
-        self.SENSOR_ID_SP2UR             = 0xC6 # Sensor ID  6
-        self.SENSOR_ID_27                = 0x1B # Sensor ID 27
+        self.SENSOR_ID_VARIO             = 0x00 # Sensor ID  0  # noqa
```
Comment (tridge) [inline Tools/autotest/common.py:837]: avoid noqa

**Review by tridge — CHANGES_REQUESTED** [review]

File: Tools/autotest/common.py (hunk 2 above)
Comment (peterbarker) [reply Tools/autotest/common.py]: Removed.

File: Tools/autotest/common.py (hunk 3 above)
Comment (peterbarker) [reply Tools/autotest/common.py]: done

File: Tools/autotest/common.py [hunk 4]
```diff
@@ -1208,13 +1253,14 @@ def progress(self, text, send_statustext=True):
...
         delta_time = time.time() - __autotest__.start_time
-        formatted_text = "AT-%06.1f: %s" % (delta_time,text)
+        formatted_text = "AT-%06.1f: %s" % (delta_time, text)
         print(formatted_text)
-        if (send_statustext and
-            self.mav is not None and
-            self.last_progress_sent_as_statustext != text):
-                self.send_statustext(formatted_text)
-                self.last_progress_sent_as_statustext = text
+        if (
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply Tools/autotest/common.py]: fixed.

**Comment by peterbarker** [comment]
```
pbarker@bluebottle:~/rc/ardupilot(pr/flake8-clean)$ ./Tools/scripts/run_flake8.py 
****** Checking (Tools/scripts/run_flake8.py)
****** Checking (Tools/gittools/pre-commit.py)
****** Checking (Tools/autotest/arduplane.py)
****** Checking (Tools/autotest/ardusub.py)
****** Checking (Tools/autotest/rover.py)
****** Checking (Tools/autotest/common.py)
****** Checking (Tools/autotest/autotest.py)
****** Checking (Tools/autotest/antennatracker.py)
pbarker@bluebottle:~/rc/ardupilot(pr/flake8-clean)$ 
```

**Comment by tridge** [comment]
happy to merge when passes CI

**Review by tridge — APPROVED** [review]

**Pull request merged**